pkg/sdk/endpointslice_controller.go

```diff
@@ -0,0 +1,102 @@
+package sdk
+
+import (
+	"context"
+	"fmt"
+
+	discoveryV1 "k8s.io/api/discovery/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctlr "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// KubernetesServiceNameIndexField is the name of the Index Field we use to index EndpointSlices by their service owners.
+	KubernetesServiceNameIndexField = "k8sServiceName"
+	// KubernetesServiceNameLabel is the label used to identify the Kubernetes service name on an EndpointSlice.
+	KubernetesServiceNameLabel = "kubernetes.io/service-name"
+)
+
+type endpointSliceReconciler struct {
+	client.Client
+	scheme *runtime.Scheme
+	impl   EndpointSliceImpl
+}
+
+// ServiceNameIndexFunc is a client.IndexerFunc that parses a Kubernetes object and returns the value of the Kubernetes service-name label.
+// Used to index EndpointSlices by their service owners.
+func ServiceNameIndexFunc(obj client.Object) []string {
+	slice, ok := obj.(*discoveryV1.EndpointSlice)
+	if !ok {
+		return nil
+	}
+
+	name := GetServiceNameFromEndpointSlice(slice)
+	if name == "" {
+		return nil
+	}
+
+	return []string{name}
+}
+
+// GetServiceNameFromEndpointSlice returns the value of the Kubernetes service-name label from an EndpointSlice.
+func GetServiceNameFromEndpointSlice(slice *discoveryV1.EndpointSlice) string {
+	if slice.Labels == nil {
+		return ""
+	}
+
+	return slice.Labels[KubernetesServiceNameLabel]
+}
+
+// RegisterEndpointSliceController registers the EndpointSliceController in the manager.
+func RegisterEndpointSliceController(mgr manager.Manager, impl EndpointSliceImpl) error {
+	r := &endpointSliceReconciler{
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+		impl:   impl,
+	}
+
+	err := mgr.GetFieldIndexer().IndexField(context.TODO(), &discoveryV1.EndpointSlice{}, KubernetesServiceNameIndexField, ServiceNameIndexFunc)
```

Comment:
No, I should just fix it now. I keep using `context.TODO()` as a placeholder and then forgetting to replace it. 

Do you think `context.WithTimeout` is reasonable here?

Reply:
context.WithTimeout sounds reasonable. However, I think it should also depend on the signal handler context, so that if Gateway pod gets shutdown, the IndexField unblocks.
```
	ctx := ctlr.SetupSignalHandler()
```

Reply:
Ah, yes, good idea. Do you think using the ctx from the controller is sufficient?

Reply:
> Do you think using the ctx from the controller is sufficient?

If we use that ctx, will there be a chance that the IndexField() hangs indefinitely? And, as a result, the pod will just hang? If that could happen, I think a timeout based on the controller ctx is a good idea.